aidermacs-models.el

```diff
@@ -9,22 +9,92 @@
-  :type '(repeat string)
-  :group 'aidermacs-models)
+(require 'json)
+(require 'url)
+
+(defun fetch-openai-compatible-models (url)
+  "Fetch available models from an OpenAI compatible API endpoint at URL."
+  (let* ((url-parsed (url-generic-parse-url url))
+         (hostname (url-host url-parsed))
+         (prefix (cond ((string= hostname "api.openai.com") "openai")
+                      ((string= hostname "openrouter.ai") "openrouter")
+                      ((string= hostname "api.deepseek.com") "deepseek")
+                      ((string= hostname "api.anthropic.com") "anthropic")
+                      ((string= hostname "generativelanguage.googleapis.com") "gemini")
+                      (t (error "Unknown API host: %s" hostname))))
+         (token (cond ((string= hostname "api.openai.com") (getenv "OPENAI_API_KEY"))
+                     ((string= hostname "openrouter.ai") (getenv "OPENROUTER_API_KEY"))
+                     ((string= hostname "api.deepseek.com") (getenv "DEEPSEEK_API_KEY"))
+                     ((string= hostname "api.anthropic.com") (getenv "ANTHROPIC_API_KEY"))
+                     ((string= hostname "generativelanguage.googleapis.com") (getenv "GEMINI_API_KEY"))
+                     (t (error "Unknown API host: %s" hostname)))))
+    (with-current-buffer
+        (let ((url-request-extra-headers
+               (cond ((string= hostname "api.anthropic.com")
+                     `(("x-api-key" . ,token)
+                       ("anthropic-version" . "2023-06-01")))
+                     ((string= hostname "generativelanguage.googleapis.com")
+                      nil)  ; No auth headers for Gemini, key is in URL
+                     (t
+                      `(("Authorization" . ,(concat "Bearer " token)))))))
+          (url-retrieve-synchronously
+           (if (string= hostname "generativelanguage.googleapis.com")
+               (concat url "/models?key=" token)
+             (concat url "/models"))))
+      (goto-char url-http-end-of-headers)
+      (let* ((json-object-type 'alist)
+             (json-data (json-read))
+             (models (if (string= hostname "generativelanguage.googleapis.com")
+                        (alist-get 'models json-data)
+                      (alist-get 'data json-data))))
+        (mapcar (lambda (model)
+                  (concat prefix "/"
+                          (cond
+                           ((string= hostname "generativelanguage.googleapis.com")
+                            (replace-regexp-in-string "^models/" "" (alist-get 'name model)))
+                           ((stringp model) model)  ; Handle case where model is just a string
+                           (t (or (alist-get 'id model)
+                                (alist-get 'name model))))))
+                models)))))
+
+(defun aidermacs--get-supported-models ()
+  "Get list of models supported by aider."
+  (with-temp-buffer
+    (call-process "aider" nil t nil "--list-models" "/")
```

Comment:
I just pushed an big update, please rebase against master.

My update introduced `aidermacs--current-output` that will store the last output from aider, it works wonderfully with `comint` and `vterm` is quite a troublesome hack. Let me know if you can improve it. 

So instead of calling a new `aider` instance, I'd prefer to use this interface and call `/models` on an existing aider session. It stores the output to `aidermacs--current-output` then you can do some regex magic on it.

Reply:
Thanks for the suggestion. That works. It does come with a caveat that if no aider is running that won't work. I added back the popular default for that case.

Reply:
> It does come with a caveat that if no aider is running that won't work. I added back the popular default for that case.

Thoughtful!